doc/factors.md

```diff
@@ -0,0 +1,218 @@
+
+### Requirements:
+
+Marian from this repository.
+
+## Define factors
+
+Factors should be organized in "groups," where each group represents a different feature. For example, there could be a group denoting capitalization and another denoting subword divisions.
+
+Factors within a single group should start with the same string.
+
+For example, for a capitalization factor group, the individual factors could be:
+
+* `c0`: all lowercase
+
+* `c1`: first character capitalized, rest lowercase
+
+* `c2`: all uppercase
+
+If there were a second factor group for subword divisions, the individual factors could be:
+
+* `s0`: end of word, whitespace should follow
+
+* `s1`: join token with next subword
+
+There is no limit on the number of factor groups barring some practical limitations having to do with how the vocabulary is stored by `marian`. If the limit is exceeded `marian` will throw this error.
+
+Factor group zero is always the actual words in the text, referred to as *lemmas*.
+
+## Data preparation
+
+Factors are appended to the *lemmas* with a pipe `|`. The pipe also separates factors of multiple groups.
+
+Example sentence:
+
+```
+Trump tested positive for COVID-19.
+```
+
+Preprocessed sentence:
+```
+trump test@@ ed positive for c@@ o@@ v@@ i@@ d - 19 .
+```
+
+Apply factors:
+```
+trump|c1|s0 test|c0|s1 ed|c0|s0 positive|c0|s0 for|c0|s0 c|c2|s1 o|c2|s1 v|c2|s1 i|c2|s1 d|c2|s0 -|c0|s0 19|c0|s0 .|c0|s0
+```
+
+
+## Create the factored vocabulary
+
+Factored vocabularies should have the extension `.fsv`. How to structure the vocabulary file is described below. If using factors only on the source or target side, the vocabulary of the other side can be a normal `json`, `yaml`, etc.
+
+The `.fsv` vocabulary must have two sections:
+
+1. **Factors**
+
+    The factor groups are defined with an underscore prepended. The colon indicates which factor group each factor inherits from. `_has_c` is used in the definition of the words in the vocabulary (see #2 below) to indicate that that word has that factor group. The `_lemma` factor is used for the words/tokens themselves; this must be present.
+
+    ```
+    _lemma
+
+    _c
+    c0 : _c
+    c1 : _c
+    c2 : _c
+    _has_c
+
+    _s
+    s0 : _s
+    s1 : _s
+    _has_s
+    ```
+
+2. **Lemmas**
+
+    These are the vocabulary entries themselves. They have the format of `LEMMA : _lemma [_has_c] [_has_s]`. The `_has_X` should only apply to lemmas that can have an `X` factor anywhere in the data (which will likely be all of the tokens except `</s>` and `<unk>`).
+
+    Examples:
+    ```
+    </s> : _lemma
+    <unk> : _lemma
+    , : _lemma _has_c _has_s
+    . : _lemma _has_c _has_s
+    the : _lemma _has_c _has_s
+    for: _lemma _has_c _has_s
+    ```
+
+
+#### Other suggestions
+
+Certain characters are used by the `.fsv` vocabulary that will have to be escaped/replaced in the data: `#:_\|`
```

Comment:
Yeah actually I also never saw this character having a critical part on the vocabulary like the other 4. We suggest to escape it because that is what we were told in one of the early meetings we had with who implemented the code in the first place. I escape the characters by replacing those special characters in the corpus before creating the vocabulary. You could do it with this `sed` line for example: 
```bash
cat corpus | sed -e 's/#/\&htg;/g' -e 's/:/\&cln;/g' -e 's/_/\&usc;/g' -e 's/|/\&ppe;/g' -e 's/\\/\&esc;/g' > corpus.esc
```

Reply:
I guess `\` could interpreted as some base64-encoded string.

It seems that `<->` is a special string too; source code gives an example of: `_x <-> _has_x`.

Reply:
As you can see here https://github.com/marian-nmt/marian-dev/pull/748#discussion_r638153012 That characters are no longer used in the code thus not needed to be escaped. Also since it is a line that starts with _ it already enters the special line treatment in the factored vocab code. So it should not be a problem if present in the lemma vocabulary

Reply:
If it's required not to break older vocabs, could you update the comment in the code to reflect the that it's no longer used.

Reply:
No, you can safely omit the `_x <-> _has_x` form the vocabulary file.